Describe layout and FP16 support in MO (#10820)
openvinotoolkit/openvino: docs/MO_DG/prepare_model/Additional_Optimizations.md

    OpenVINO Runtime API [ShapeInference feature](../../OV_Runtime_UG/ShapeInference.md).

## When to Specify Layout

Layout defines which dimensions are located in shape and can be specified
for both inputs and outputs. For layout syntax check the [Layout API overview](../../OV_Runtime_UG/layout_overview.md). 
To specify layout, you can use either `--layout` or `--source_layout` followed by the layout value. 
Additionally, if a model has more than one input or needs both input and output layouts specified,
each layout needs to be given a name. Here are examples of the two cases:
`NCHW` layout specified for a model with a single input:
```
mo --input_model /path/to/model --source_layout nchw
mo --input_model /path/to/model --layout nchw
```
`NCHW` and `NC` layouts specified for a model with two inputs:
```
mo --input_model /path/to/model --source_layout name1(nchw),name2(nc)
mo --input_model /path/to/model --layout name1(nchw),name2(nc)
```
Some preprocessing may require setting of input layouts, for example: batch setting,
application of mean or scales, and reversing input channels (BGR<->RGB).

## How to Change Layout of a Model 

> It is possible to change layout of the model. It may be needed if input data has different
> layout then model was trained on. There are 2 options that can be used to change layout of
> model inputs or outputs --layout and --target_layout. For example to change layout of the
> model with one input from `NHWC` to `NCHW`:
> ```
> mo --input_model /path/to/model --source_layout nhwc --target_layout nchw
> mo --input_model /path/to/model --layout "nhwc->nchw"
> ```
> Similarly, if model has multiple inputs or if it is required to change layout of output,
> names must be specified:
> ```
> mo --input_model /path/to/model --source_layout name1(nhwc),name2(nc) --target_layout name1(nchw)
> mo --input_model /path/to/model --layout "name1(nhwc->nchw),name2(nc)"
> ```

Reviewer: ```suggestion
Changing model layout may be necessary if it differs from the one presented by input data. 
Similarly to specifying layout, to change it using one of two commands: either `--layout` or `--target_layout`. 
Again, if a model has more than one input or needs both input and output layouts specified,
each layout needs to be given a name. Here are examples of the two cases:

`NHWC` layout of a single-input model changed to `NCHW`:
```
mo --input_model /path/to/model --source_layout nhwc --target_layout nchw
mo --input_model /path/to/model --layout "nhwc->nchw"
```

`NHWC` layout of a multi-input model changed to `NCHW`:
```
mo --input_model /path/to/model --source_layout name1(nhwc),name2(nc) --target_layout name1(nchw)
mo --input_model /path/to/model --layout "name1(nhwc->nchw),name2(nc)"
```
```

-	**Larger batch size**<br>
	Notice that the devices like GPU are doing better with larger batch size. While it is possible to set the batch size in the runtime using the OpenVINO Runtime API [ShapeInference feature](../../OV_Runtime_UG/ShapeInference.md).

> ## When to Specify Layout

Reviewer 2: In this document we also have a mean / scale options which are also require layouts to be specified. So, should not descriptions of layouts be before other options?

Author: It seems correct order to me: firstly we list the optimizations that can be done, secondly, we describe when layouts are needed and after that we describe each optimization.

to be specified it is required to specify names for each layout:
```
mo --input_model /path/to/model --source_layout name1(nchw),name2(nc)
mo --input_model /path/to/model --layout name1(nchw),name2(nc)
```
Some preprocessing may require setting of input layouts, for example setting batch,
application of mean or scales, reverse input channels (BGR<->RGB)

## How to change layout of the model 

It is possible to change layout of the model. It may be needed if input data has different
layout then model was trained on. There are 2 options that can be used to change layout of
model inputs or outputs --layout and --target_layout. For example to change layout of the
model with one input from NHWC to NCHW:
```
mo --input_model /path/to/model --source_layout nhwc --target_layout nchw
mo --input_model /path/to/model --layout "nhwc->nchw"
```
Similarly, if model has multiple inputs or if it is required to change layout of output,
names must be specified:
```
mo --input_model /path/to/model --source_layout name1(nhwc),name2(nc) --target_layout name1(nchw)
mo --input_model /path/to/model --layout "name1(nhwc->nchw),name2(nc)"
```

## When to Specify Mean and Scale Values

> Usually neural network models are trained with the normalized input data. This means that the input data values are converted to be in a specific range, for example, `[0, 1]` or `[-1, 1]`. Sometimes the mean values (mean images) are subtracted from the input data values as part of the pre-processing. There are two cases how the input data pre-processing is implemented.

Author: You finished this paragraph with `:` like if there will be a list after it while it should end with `.`

> Usually neural network models are trained with the normalized input data. This means that the input data values are converted to be in a specific range, for example, `[0, 1]` or `[-1, 1]`. Sometimes the mean values (mean images) are subtracted from the input data values as part of the pre-processing. There are two cases how the input data pre-processing is implemented.

Reviewer: I did it due to the two points "part of a topo" and "not part of a topo"
If we want to merge the following "in the first / second case" parts with the points, I totally agree, we can go with a full stop.

> Usually neural network models are trained with the normalized input data. This means that the input data values are converted to be in a specific range, for example, `[0, 1]` or `[-1, 1]`. Sometimes the mean values (mean images) are subtracted from the input data values as part of the pre-processing. There are two cases how the input data pre-processing is implemented.

Author: Excuse me, I didn't notice that there is a continuation of this paragraph

## When to Reverse Input Channels <a name="when_to_reverse_input_channels"></a>
Input data for your application can be of RGB or BRG color input order. For example, OpenVINO Samples load input images in the BGR channels order. However, the model may be trained on images loaded with the opposite order (for example, most TensorFlow\* models are trained with images in RGB order). In this case, inference results using the OpenVINO samples may be incorrect. The solution is to provide `--reverse_input_channels` command line parameter. Taking this parameter, the Model Optimizer performs first convolution or other channel dependent operation weights modification so these operations output will be like the image is passed with RGB channels order.

## Compression of model to FP16

> Mode Optimizer can compress the model to `FP16` data type. This will not only make the
> model to occupy less space in file system, but also may increase performance on some
> hardware. Model Optimizer will change data type on all the constants inside the model
> to `FP16` precision and insert `Convert` nodes to initial data type so that the data
> flow inside the model is preserved. To compress the model to `FP16` please use
> `--data_type` option like this:

Reviewer: ```suggestion
Model Optimizer can compress models to `FP16` data type. This makes them occupy less space 
in the file system and, most importantly, increase performance when particular hardware is used. 
The process assumes changing data type on all constants inside the model
to the `FP16` precision and inserting `Convert` nodes to the initial data type, so that the data
flow inside the model is preserved. To compress the model to `FP16` use the `--data_type` option like this:
```

```
mo --input_model /path/to/model --data_type FP16
```

> > **NOTE**: Using `--data_type FP32` will not do anything and will not force `FP32` 
> > precision in the model. If the model was `FP16` originally in the framework,
> > Model Optimizer will not convert such weights to `FP32` even if `--data_type FP32`
> > option is used .

Reviewer: ```suggestion
> **NOTE**: Using `--data_type FP32` is not a proper command. It will not force conversion of an `FP16`model 
> to the precision of `FP32`. 
```

Reviewer: Yep, I was thinking about it a lot... I decided to use "not a proper command" to be on the safe side, but you might be right, maybe we should make it even safer... Maybe we could go with "Using --data_type FP32 will give no result. It will not..."

Author: I did the different wording, please review it.

> Some plugins, for example GPU, will show greater performance while slightly sacrificing

Reviewer 2: greater => better?
@vladimir-paramuzov do we need to write about GPU here? Maybe we just need to refer to GPU paragraph with precisions description?

Reviewer: Although they are often synonymous, we can change "greater" to "higher". But more importantly, I agree, this line is not necessary, as it is mentioned more or less in the beginning paragraph. We could just mention GPU there. And the following line should also go to the top, as it is another reason for compressing, just like the size and performance boost.
